sirepo/pkcli/runner.py

```diff
@@ -0,0 +1,366 @@
+    for k in list(env):
+        if _EXEC_ENV_REMOVE.search(k):
+            del env[k]
+    env['SIREPO_MPI_CORES'] = str(mpi.cfg.cores)
+    return env
+
+
+class _JobInfo:
+    def __init__(self, run_dir, jhash, status, process):
+        self.run_dir = run_dir
+        self.jhash = jhash
+        self.status = status
+        self.finished = trio.Event()
+        self.process = process
+
+
+class _JobTracker:
+    def __init__(self, nursery):
+        # XX TODO: eventually we'll need a way to stop this growing without
+        # bound, perhaps by clarifying the split in responsibilities between
+        # the on-disk simulation_db versus the in-memory status.
+        self.jobs = {}
+        self._nursery = nursery
+
+    def status(self, run_dir, jhash):
+        disk_in_path = run_dir.join('in.json')
+        disk_status_path = run_dir.join('status')
+        if disk_in_path.exists() and disk_status_path.exists():
+            disk_in_text = pkio.read_text(disk_in_path)
+            disk_jhash = pkjson.load_any(disk_in_text).reportParametersHash
+            if disk_jhash == jhash:
+                disk_status = pkio.read_text(disk_status_path)
+                if disk_status == 'pending':
+                    # We never write this, so it must be stale, in which case
+                    # the job is no longer pending...
+                    return runner_client.JobStatus.ERROR
+                return runner_client.JobStatus(disk_status)
+        if run_dir in self.jobs and self.jobs[run_dir].jhash == jhash:
+            return self.jobs[run_dir].status
+        return runner_client.JobStatus.MISSING
+
+    async def start_job(self, run_dir, jhash, cmd):
+        await self._nursery.start(self._run_job, run_dir, jhash, cmd)
+
+    async def _run_job(
+            self, run_dir, jhash, cmd, *, task_status=trio.TASK_STATUS_IGNORED
+    ):
+        # XX TODO: there are all kinds of race conditions here, e.g. if the
+        # same jid gets started multiple times in parallel... we need to
+        # revisit this once jids are globally unique, and possibly add
+        # features like "if there is another job running with the same
+        # user/sim/job but a different hash, then auto-cancel that other job"
+        with _catch_and_log_errors(Exception, 'error in run_job'):
+            if run_dir in self.jobs:
+                # Right now, I don't know what happens if we reach here while
+                # the previous job is still running. The old job might be
+                # writing to the new job's freshly-initialized run_dir? This
+                # will be fixed once we move away from having server.py write
+                # directly into the run_dir.
+                assert self.jobs[jhash] == jhash
+                return
+            try:
+                env = _subprocess_env()
+                run_log_path = run_dir.join(template_common.RUN_LOG)
+                # hack: switch back to py2 mode to run the actual command
```

Comment:
Not a hack. It's going to be what we do for a while. It will eventually need to be encapsulated, because the docker container will need the same code.

Reply:
the hacky bit is that py2 is normally the default, but we're in this special py3 environment so we have to take some special effort to prevent processes we spawn from inheriting that. docker won't have the problem, because containers won't inherit the runner's environment.